### DISCUSSION TITLE
Adding NGINX infront of TURN

### DISCUSSION THREAD

**Discussion by WinstonPais** [coturn/coturn#1160] [Q&A]
I am trying to add NGINX infront of my TURN server

My NGINX conf looks like this.
```
stream {
     upstream turn
        {
                server 127.0.0.1:5349;
        }
     server {
                listen 443 udp;

                resolver 1.1.1.1;
                proxy_connect_timeout 5s;
                proxy_timeout 15s;
                proxy_protocol on;

                proxy_pass turn;
                ssl_preread on;
     }
}
```
This works on the Chrome browser. I get srflx and relay candidates.

But on Firefox get none of these candidates.

Even on chrome there is a slight difference in the candidates I get with and without NGINX which could be the cause.

Without NGINX:
```
candidate:3646143538 1 udp 1677729535 <my_ip> 59271 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag xmIu network-cost 999

candidate:1499094430 1 udp 33562623 <public_ip_of_my_turn> 63683 typ relay raddr <my_ip> rport 59271 generation 0 ufrag xmIu network-cost 999
```
With NGINX:
```
candidate:4227863252 1 udp 1677729535 127.0.0.1 54974 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag 0nJM network-cost 999

candidate:601945334 1 udp 33562623 <public_ip_of_my_turn> 60590 typ relay raddr 127.0.0.1 rport 54974 generation 0 ufrag 0nJM network-cost 999
```
As we can see i get raddr as localhost. Passing the remote IP properly to the backend could solve the issue.

Even in coturn logs: without NGINX: `remote <my_ip>:59271`

with NGINX: `remote 127.0.0.1:54974`

**Comment by eakraly** [comment]
Hi @WinstonPais 

I'm not that familiar with NGINX usage but I may still have an idea (regarding 127.0.0.1 which is definitely incorrect).

It seems like the issue is not related to NGINX but the network: turnserver selects relay interface based on the IP of the client connection. In your setup, NGINX forwards the connection using 127.0.0.1 - and turnserver decides to use localhost because that's where the connection was receive from. See related comment in the configuration example https://github.com/coturn/coturn/blob/master/examples/etc/turnserver.conf#L99

```
# If no relay IP(s) specified, then the turnserver will apply the default
# policy: it will decide itself which relay addresses to be used, and it
# will always be using the client socket IP address as the relay IP address
# of the TURN session (if the requested relay address family is the same
# as the family of the client socket).
```

I think that if you run turnserver with `--relay-ip=<my_ip>` it will fix the issue (allocations will be using <my_ip> instead of localhost)
